hazelcast/src/main/java/com/hazelcast/client/impl/ClientEngineImpl.java

```diff
@@ -329,14 +328,42 @@ public SecurityContext getSecurityContext() {
     }
 
     @Override
-    public void bind(final ClientEndpoint endpoint) {
-        final Connection conn = endpoint.getConnection();
-        if (conn instanceof TcpIpConnection) {
-            InetSocketAddress socketAddress = conn.getRemoteSocketAddress();
-            //socket address can be null if connection closed before bind
-            if (socketAddress != null) {
-                Address address = new Address(socketAddress);
-                ((TcpIpConnection) conn).setEndPoint(address);
+    public boolean bind(final ClientEndpoint endpoint) {
+        if (!clientSelector.select(endpoint)) {
+            return false;
+        }
+        if (endpointManager.registerEndpoint(endpoint)) {
+            Connection conn = endpoint.getConnection();
+            if (conn instanceof TcpIpConnection) {
+                InetSocketAddress socketAddress = conn.getRemoteSocketAddress();
+                //socket address can be null if connection closed before bind
+                if (socketAddress != null) {
+                    Address address = new Address(socketAddress);
+                    ((TcpIpConnection) conn).setEndPoint(address);
+                }
+            }
+        } else {
+            // connection exists just reauthenticated to promote to become an owner connection
```

Comment:
minor: this might be more readable if you move this condition above like;

```java
if (!endpointManager.registerEndpoint(endpoint)) {
   // connection exists just reauthenticated to promote to become an owner connection
   return true;
}
```